Improve the accuracy of the PQ docs.

jordansissel opened the pull request:
* Removed reference to unused `queue.checkpoint.interval`.
* Changed 'unread' to be 'unACKed'
* Remove confusing/contradictory statement about queue acknowledgement
  and outputs

colinsurprenant commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,47 +1,45 @@
-shutdown, it's possible to lose queued events. 
+(inputs → pipeline workers) to buffer events. The size of these in-memory
+queues is fixed and not configurable. If Logstash terminates abnormally,
+either as the result of a software crash of  `kill -9`, it's
```
of -> or ?

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,47 +1,45 @@
-sent to the <<plugins-inputs-beats,beats>> input. For other inputs, Logstash
-only acknowledges delivery of messages in the filter and output stages, and not
-all the way back to the input or source.
+* Inputs which do not use a request-response protocol cannot be protected from data loss. For example: tcp, udp, zeromq push+pull, and many other inputs do not have a mechanism to acknowledge receipt to the sender. Plugins such as beats and http, which *do* have a acknowledgement capability, are well protected by this queue.
```
Not sure about my wording choice here. Intent was to highlight that protection is only viable for protocols that have a reply mechanism for data. Protocols with no reply mechanism (udp, tcp, etc) would still be queued but we have no way to report back to the sender.

dedemorton replied:
Wording seems clear to me. I would say "Inputs that do not use..." (use "that" for restrictive clauses).

jordansissel replied:
:heart:

jordansissel replied:
@acchen97 I can't tell if the Kafka input actually does this. https://github.com/logstash-plugins/logstash-input-kafka/blob/master/lib/logstash/inputs/kafka.rb#L245-L260 shows no indication that I can see that it does do this acking, but it may do it somewhere deeper inside the Kafka client.

acchen97 replied:
@jordansissel I spoke to @talevy and Kafka input does ack.

In this context, I think we can just note Beats, Kafka, RabbitMQ, and HTTP inputs.

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -53,44 +51,46 @@ process:
+are processed, Logstash records a checkpoint to track which events are
+successfully acknowledged (ACKed) as processed by Logstash. An event is
+recorded as ACKed in the checkpoint if the event is successfully sent to the
+last output stage in the pipeline;
```
This should be a period

jordansissel replied:
But semicolons are my favorite!

Thanks :)

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -53,44 +51,46 @@ process:
 of these events, and then terminates the Logstash process. Upon restart,
-Logstash uses the checkpoint file to pick up where it left off in the persistent
-queue and processes the events in the backlog. 
+Logstash uses last checkpoint to resume processing the events in the backlog. 
```
Uses the last checkpoint

colinsurprenant replied:
there is not really a "last" checkpoint - there are always 1 checkpoint per page. Do we need to talk about checkpoints here? maybe just saying that logstash will resume processing where it left?

jordansissel replied:
Yeah, the doc had 'checkpoint file' in various places, and I mostly patched this to remove the word 'file' (because that's an implementation detail unimportant to most users)

I think in most case we are trying to convey the concept that a checkpoint can be behind real-time and that the consequence of this is that an abnormal termination will cause data loss for all data written after the most recent checkpoint ("checkpoint" as an action, not a file).

jordansissel replied:
In places where we are describing order of operations, I am wondering if it would be more clear to have a diagram instead of a paragraph of text.

dedemorton replied:
👍

jordansissel replied:
@colinsurprenant I am also with you on maybe not needing to discuss checkpoints in the way this document does today -- I think we should mention it with respect to durability and maybe nowhere else. At present, it's spread around the document and maybe needs to be just in one section.

colinsurprenant replied:
durability++

colinsurprenant commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -53,44 +51,46 @@ process:
+persistent queue and sends them to the filter and output stages. As events
+are processed, Logstash records a checkpoint to track which events are
+successfully acknowledged (ACKed) as processed by Logstash. An event is
+recorded as ACKed in the checkpoint if the event is successfully sent to the
```
«recorded as ACKed in the checkpoint» I would not phrase like this because we actually don't record acking per event in the checkpont - we keep track of the first unacked event in a page . I would just remove "in the checkpoint" - specific ACK tracking IS done in-memory during runtime but not in the checkpoint.

colinsurprenant commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -53,44 +51,46 @@ process:
 * `path.queue`: The directory path where the data files will be stored. By default, the files are stored in `path.data/queue`. 
 * `queue.page_capacity`: The size of the page data file. The queue data consists of append-only data files separated into pages. The default size is 250mb. 
-* `queue.max_events`:  The maximum number of unread events that are allowed in the queue. The default is 0 (unlimited).
+* `queue.max_events`:  The maximum number of unACKed events that are allowed in the queue. The default is 0 (unlimited).
```
no, it really *is* the max number of unread events.

jordansissel replied:
In this doc there is no other mention of the word "unread". If we're going to mention it now, it probably needs explanation. It sounds like there's multiple states (read, unread, acked, unacked?) for an event, and that is news to me, can you help me understand more?

colinsurprenant replied:
sure, a queue event can be in the following state:
- unread (which also implies unacked)
- read & unacked (thus "in flight" in the worker) 
- read & acked
Now, read & acked events only exists because purging is done at page level. only when all events within a page are acked is it purged.

jordansissel commented:
I'm still working on this. Based on feedback and other discussion, I'm going to do some more possibly-larger changes to this doc in this PR. Will update when it's ready for the next round of review.

Thank you all for the feedback so far!

jordansissel commented:
Ok I did another pass at the content in this document.

I've tried to do the following:

* Focus durability discussion only on the durability section
* Focus on 'ack' terminology
* consistent use of phrase "abnormal termination" when discussing that topic
* Remove "Advantages of Persistent Queue" section since it's just a bullet-point summary of the first section. Moved it into the first section.
* Defined "acknowledge" for pipeline workers, sorta.
* Added caveat on `queue.max_events` as being not intended for user benefit.
* Added necessary background for durability discussion (pages, checkpoints)
* Described garbage collection scenario of fully acked pages.

jordansissel commented:
@colin @dedemorton @acchen97 Would love your feedback on the current state.

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+the write to the queue is successful, the input can send an acknowledgement to
+its data source.
+
+When processing events from the queue, events are only acknowledged as
```
Sentence starts with a dangling modifier (a phrase that modifies a word that isn't clearly stated in the sentence). How about:

When processing events from the queue, Logstash acknowledges events as completed, within the queue, only after the filters and outputs have completed.

[changed to "after" because "once" is easily misread/mistranslated as "one time"]

jordansissel replied:
++

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+
+When processing events from the queue, events are only acknowledged as
+completed, within the queue, once the filters and outputs have completed.
+The queue keeps a record of events which have been processed by the pipeline.
```
"keeps a record of events that have been processed..."

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+"ACKed") if, and only if, the event has been processed completely by the
+Logstash pipeline. 
+
+What does acknowledged mean? For the common case, this means that it has been
```
I am unsure what you  mean by "common case" here. Do you mean "in most cases"? Also the antecedent for "it" is unclear. Does "it" refer back to the event? If so, how about:

"What does "acknowledged" mean? In most cases, this means that the event has been handled by all configured filters and outputs."

jordansissel replied:
++

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+
+What does acknowledged mean? For the common case, this means that it has been
+handled by all configured filters and outputs. For example, if you have only
+one output, Elasticsearch, an event is only ACKed once the Elasticsearch output
```
...an event is ACKed when the Elasticsearch output has successfully sent this event to Elasticsearch... [again, "once" is easily misread/mistranslated at "one time"]

jordansissel replied:
++

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -101,23 +103,24 @@ queue.max_bytes: 4gb
-flow at the input stage without overwhelming downstream stages and outputs like
+When the queue is full, Logstash puts back pressure on the inputs to stall data
+flowing into Logstash. This mechanism helps Logstash control the rate of data
+flow at the input stage without overwhelming outputs and outputs like
```
You say outputs twice. Maybe you mean to say:

 "...without overwhelming outputs like  Elasticsearch."

jordansissel replied:
++

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -101,23 +103,24 @@ queue.max_bytes: 4gb
+size of the queue reaches 8gb. When the queue is full of unACKed events, and
 the size limit has been reached, Logstash will no longer accept new events. 
 
+On disk, the queue is stored as a set of pages where each page is one file. Each page can be at most `queue.page_capacity` in size. Pages are deleted (garbage collected) once all events in that page have been ACKed. If an older page has at least one event that is not yet ACKed, that entire page will remain on disk until all events in that page are successfully processed. Each page containing unpr ...
```
I wish we didn't have to reveal these details because they seem like internals that users shouldn't need to know about. This paragraph seems like a bit of a non-sequitur. Maybe you just need to add a line that explains how the setting affects back pressure (it's a bit murky to me, TBH).

Also change: "Pages are deleted (garbage collected) **once** " to "Pages are deleted (garbage collected) **after** all events..."

acchen97 replied:
I agree with @dedemorton a bit here in that I think this level of detail would be better for a blog post.

jordansissel replied:
I would love to avoid writing details about the `page` concept. I don't know how to do this, though, given the behavior of the queue.

The problem is this: A page with *any* unacked events will remain completely on disk. This can mean that for a 250mb page, if only 1KB is unacked, then we will not clean it up. The impact of this capacity planning and behavior for really small queue (byte) sizes, I believe, can become very confusing.

For my concern here, it is mostly predicted user confusion and my strong desire to not omit information that causes users to file bug reports or customers to file support tickets.

acchen97 replied:
@jordansissel I'm curious in what scenarios this would occur. If all events in a page are successful except for one, won't that event continue to retry for a number of times until it's dropped (and therefore ack'ed)? I feel like we'd be able to solve this altogether when we have DLQs, or introducing something like compaction though that may be overkill.

It sounds like you're positioning this as an unexpected caveat/gotcha to when backpressure may be exerted. Maybe end with "Therefore, if there are unprocessed events in the queue, the 'queue.max_bytes' limit may be hit unexpectedly, exerting backpressure back to the source." Without a mechanism for queue eviction, can users work around this today aside from bumping up their queue.max_bytes?

colinsurprenant replied:
the situation where a page is not deleted because all events are not yet acked can happen in these 2 scenarios:
1- outputs are slower than inputs, which will typically result in the accumulation of multiple pages and all events in the page(s) will be unacked until the outputs are ready to process a page. Typically there will be a short overlap when some output start processing a new page and other output finish processing the previous page.
2- an output (or multiple outputs) are "stuck" in a retry loop for some time and some events will remain unacked during this retry period so a page might be kept "alive" (not deleted) because of this particular output "stuck" in retry for some time while other output might actually finish processing the rest of the page and move on to the other page. This situation will be temporary up to be output retry thresholds which is plugin dependant.

jordansissel replied:
I've moved this part about pages and garbage collection into a new section called 'disk garbage collection'

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,22 +131,35 @@ events.
-durability by providing the following configuration options:
+disk. Logstash commits to disk in a mechanism we call checkpointing.
+
+To discuss durability, we need to introduce a few details about how the persistent queue is implemented.
```
Hmmm....we're sort of started down this road earlier. I wonder if all the info about pages could be described in one section? Only the most dedicated readers are going to want to plow through these details.

jordansissel replied:
Yeah, maybe for `pages` I can have a section dedicated to have data is purged from the queue?

jordansissel replied:
I created a separate section for page/garbage details.

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,22 +131,35 @@ events.
+
+Second, the queue tracks details about the queue (pages, acknowledgements, etc) in a separate file called a checkpoint file.
+
+When Logstash does checkpoint, the following will occur:
```
"When Logstash does checkpoint" sounds a bit off. Maybe "When Logstash records a checkpoint" or "When Logstash does checkpointing".

jordansissel replied:
How about "When checkpointing, Logstash does the following:"?

jordansissel replied:
Changed to this:

```
 When recording a checkpoint, Logstash will:

 * Call fsync on the head page.
 * Atomically write to disk the current state of the queue.
```

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,22 +131,35 @@ events.
-ensure that all messages in Logstash's queue are durable, you can set
-`queue.checkpoint.writes: 1`. However, this setting can severely impact
-performance.
+Disk writes are not free have a cost. Tuning these values higher or lower will trade durability for performance. For instance, if you want to ensure that all events from inputs queue are durable, you can set
```
Something is missing or wasn't deleted here: "Disk writes are not free have a cost"

dedemorton replied:
I think there is a word missing here (or one that should have been deleted): "For instance, if you want to ensure that all events from inputs queue are durable"

jordansissel replied:
++ yeah I didn't clean up my sentence after editing <3

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,22 +131,35 @@ events.
+
+The following settings are available to let you tune durability:
+
+* `queue.checkpoint.writes`: The number of writes from inputs after which a checkpoint is written. 
```
This description is a bit confusing because of the sentence structure. Does this say what you meant to say?

"The number of times an input writes to the queue before a checkpoint is written?"

jordansissel replied:
It's confusing because the property of the queue is for all writes, not just from a single input. And internally, it's called a "write" not an "event" which makes this even harder to document.

I really want to say this, "The number of events written to the queue before a checkpoint is written"

But ...

* It's not the number of events, it's number of writes. It just happens that, today, one write is one event.
* and, "before a checkpoint" makes it sound (incorrectly) like that we require *at least* that number of writes before Logstash will write a checkpoint, when really it's a minimum of (number of writes, number of acks) -- whichever happens first.

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,22 +131,35 @@ events.
+The following settings are available to let you tune durability:
+
+* `queue.checkpoint.writes`: The number of writes from inputs after which a checkpoint is written. 
+* `queue.checkpoint.acks`: The number of ACKs to the queue after which a checkpoint is written. This configuration controls the durability at the processing (filter + output) part of Logstash.
```
Does this work?

"The number of ACKs sent to the queue before a checkpoint is written."

colinsurprenant replied:
or "the number or events acknowledged (ACK'ed)  before a checkpoint is written" ?

dedemorton commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,22 +131,35 @@ events.
+`queue.checkpoint.writes: 1`.
 
-* `queue.checkpoint.acks`: The number of ACKs to the queue to trigger an fsync to disk. This configuration controls the durability from the consumer side.
+In Logstash 5.0 and 5.1, a `write` to the queue is counted for every event produced by an input.
```
Maybe would be better for info maintainability to say, "Starting with Logstash 5.0...." (Unless you know this behavior is changing after 5.1).

jordansissel replied:
I don't know when this behavior will change, but it's not guaranteed to always be 1:1 and I'm trying (perhaps poorly) to solve this future problem:

* We change the "one write is one event" in the code
* but, we don't change the docs
* and, then a user files a bug about the behavior
* and, then I get annoyed we didn't update the docs.

acchen97 commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,22 +131,35 @@ events.
+
+To discuss durability, we need to introduce a few details about how the persistent queue is implemented.
+
+First, the queue itself is a set of pages. There are two kinds of pages: head page and tail page. The head page is where new events are written. When this head page is of a certain size (see `queue.page_capacity), the becomes a tail page, and a new head page is created. Tail pages are immutable. 
```
This detail I feel should also be part of a blog post where we talk about PQ implementation architecture and introduce this terminology.

acchen97 replied:
We can probably leave it out of the docs.

jordansissel replied:
Pages, I believe, are important, because they matter when Logstash considers which data can be deleted from disk, and I'm trying to document this in enough detail that user questions can be just answered by pointing at the docs.

acchen97 replied:
Sounds good. Typo here "*it* becomes a tail page"

acchen97 commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -5,72 +5,73 @@ WARNING: This functionality is in beta and is subject to change. Deployment in p
-delivery succeeds at least once. In other words, messages stored in the
-persistent queue may be duplicated, but not lost.
+* Provides protection from in-flight message loss when the Logstash process is abnormally terminated.
+* Absorb bursts of events without needing an external queueing mechanism like Redis or Apache Kafka.
```
I think it's valuable for us to still outline the "at-least-once" messaging here. Maybe another point like this:

"Guarantees at-least-once delivery when used with inputs that support acknowledgements."

jordansissel replied:
In 5.1.1 it's not guaranteed because we:

1) `checkpoint.writes: 1` is necessary for such a guarantee and it is not the default.
2) and, we don't fsync the checkpoint file (a bug fixed by #6430)

jordansissel commented:
At-least-once isn't quite available yet but is the intent. We found a bug
where the checkpoint is not written to underlying hardware and Colin is (or
has already) working on a fix.

I won't document a detail we can't deliver :)

On Wed, Dec 21, 2016 at 11:41 PM Alvin Chen <notifications@github.com>
wrote:

> *@acchen97* commented on this pull request.
>
>
>
>
> ------------------------------
>
>
>
>
> In docs/static/persistent-queues.asciidoc
> <https://github.com/elastic/logstash/pull/6408#pullrequestreview-14107559>
> :
>
>
> >
>
> -Using persistent queues:
>
> -
>
> -* Provides protection from losing in-flight messages when the Logstash process is abnormally terminated.
>
> -* Absorb bursts of events without needing an external queueing mechanism like Redis or Kafka.
>
> -* Depending on configuration, can provide an at-least-once message
>
> -delivery guarantee. If Logstash is restarted while events are in-flight,
>
> -Logstash will attempt to deliver messages stored in the persistent queue until
>
> -delivery succeeds at least once. In other words, messages stored in the
>
> -persistent queue may be duplicated, but not lost.
>
> +* Provides protection from in-flight message loss when the Logstash process is abnormally terminated.
>
> +* Absorb bursts of events without needing an external queueing mechanism like Redis or Apache Kafka.
>
>
>
> I think it's valuable for us to still outline the "at-least-once"
> messaging here. Maybe another point like this:
>
>
> "Guarantees at-least-once delivery when used with inputs that support
> acknowledgements."
>
>
>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/logstash/pull/6408#pullrequestreview-14107559>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIC6nbRQNa0Bbu4B0ICQoh5YZFD3VPGks5rKimUgaJpZM4LMBEZ>
> .
>

acchen97 commented:
@jordansissel thanks for the update, I agree. Let's add back at-least-once after the bug fix and optimistic recovery.

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
-To prevent event loss in these scenarios, you can configure Logstash to use
-persistent queues. With persistent queues enabled, Logstash persists buffered
-events to disk instead of storing them in memory. 
+Temporary machine failures are scenarios where Logstash or its host machine are
```
I agree.

I've reorganized these two paragraphs to try and have a better flow to it.

colinsurprenant commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+Instead of deploying and managing a message broker, such as Redis, RabbitMQ, or
+Apache Kafka, to facilitate a buffered publish-subscriber model, you can enable
+persistent queues to buffer events on disk and remove the message broker. This
+benefits you by removing one more piece of machinery from your deployment.
```
I'm not sure I like the wording "removing one more piece of machinery"  - it feels like a message broker necessarily involves extra hardware (machinery?) but it might not ... 
Anyhow it may just be a language thing so please disregard my comment if ya'll are comfortable with this :D

jordansissel replied:
Noted. When I reread this paragraph, I think the last sentence is unnecessary anyway so I will remove it. Thanks <3

colinsurprenant commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+one output, Elasticsearch, an event ACKed when the Elasticsearch output has
+successfully sent this event to Elasticsearch. In other cases, such as
+with the `drop` filter, an event is ACKed when the `drop` filter cancels the
+event.
```
«an event is ACKed when the `drop` filter cancels the event»  this is not true. all events in a batch are acked once the batch is done being processed by the filters+outputs worker, thus a cancelled event will actually be acked at the same time as all the other events in the batch.

jordansissel replied:
Agreed.

colinsurprenant commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
-* `queue.page_capacity`: The size of the page data file. The queue data consists of append-only data files separated into pages. The default size is 250mb. 
-* `queue.max_events`:  The maximum number of unread events that are allowed in the queue. The default is 0 (unlimited).
+* `queue.page_capacity`: The maximum size of a queue page in bytes. The queue data consists of append-only files called a "pages". The default size is 250mb. Changing this value is unlikely to have performance benefits.
+// Technically, I know, this isn't "maximum number of events" it's really maximum number of events not yet read by the pipeline worker. We only use this for testing and users generally shouldn't be setting this.
```
«append-only files called a "pages"» -> append-only files called "pages"  ?

jordansissel replied:
Oops. Good catch.

colinsurprenant commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,24 +128,41 @@ events.
+[[garbage-collection]]
+==== Disk Garbage Collection
+
+On disk, the queue is stored as a set of pages where each page is one ile. Each page can be at most `queue.page_capacity` in size. Pages are deleted (garbage collected) once all events in that page have been ACKed. If an older page has at least one event that is not yet ACKed, that entire page will remain on disk until all events in that page are successfully processed. Each page containing unpro ...
```
one ile -> one file

jordansissel replied:
++

jordansissel replied:
++

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+
+In order to protect against data loss during abnormal termination, Logstash has
+a persistent queue feature which will store the message queue on disk.
+Persistent queues provides durability of data within Logstash.
```
++

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+a persistent queue feature which will store the message queue on disk.
+Persistent queues provides durability of data within Logstash.
+
+Persistent queues are also useful for Logstash deployments that need large buffers  
```
++

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+persistent queues to buffer events on disk and remove the message broker. This
+benefits you by removing one more piece of machinery from your deployment.
+
+In summary, the two benefits of enabling the persistent queue are as follows:
```
++

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+In summary, the two benefits of enabling the persistent queue are as follows:
+
+* Provides protection from in-flight message loss when the Logstash process is abnormally terminated.
+* Absorb bursts of events without needing an external buffering mechanism like Redis or Apache Kafka.
```
++

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+"ACKed") if, and only if, the event has been processed completely by the
+Logstash pipeline. 
+
+What does acknowledged mean? For most cases, this means that it has been
```
Ahh, great point. Thank you :)

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -1,96 +1,98 @@
+
+What does acknowledged mean? For most cases, this means that it has been
+handled by all configured filters and outputs. For example, if you have only
+one output, Elasticsearch, an event ACKed when the Elasticsearch output has
```
++

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,24 +128,41 @@ events.
 ==== Controlling Durability
 
+// I'm not sure about this particular description of durability. Might need some help.
+Durability is a property of storage writes that indicate the availability of
```
Thank you! I wasn't sure how to word that.

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,24 +128,41 @@ events.
 When the persistent queue feature is enabled, Logstash will store events on
-disk. The persistent queue exposes the trade-off between performance and
-durability by providing the following configuration options:
+disk. Logstash commits to disk in a mechanism we call checkpointing.
```
++

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,24 +128,41 @@ events.
+
+To discuss durability, we need to introduce a few details about how the persistent queue is implemented.
+
+First, the queue itself is a set of pages. There are two kinds of pages: head page and tail page. The head page is where new events are written. When this head page is of a certain size (see `queue.page_capacity), the becomes a tail page, and a new head page is created. Tail pages are immutable, and the head page is append-only.. 
```
++ to all.

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,24 +128,41 @@ events.
+
+First, the queue itself is a set of pages. There are two kinds of pages: head page and tail page. The head page is where new events are written. When this head page is of a certain size (see `queue.page_capacity), the becomes a tail page, and a new head page is created. Tail pages are immutable, and the head page is append-only.. 
+
+Second, the queue tracks details about the queue (pages, acknowledgements, etc) in a separate file called a checkpoint file.
```
++

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,24 +128,41 @@ events.
 
-The process of checkpointing is atomic, which means any update to the file is
-saved if successful.
+* `queue.checkpoint.writes`: Logstash will checkpoint after this many writes into the queue. At this time, one event counts as one write, but this may change in future releases.
```
Your explanations are delightfully informative. Thank you!

jordansissel commented on docs/static/persistent-queues.asciidoc:
```diff
@@ -128,24 +128,41 @@ events.
+* `queue.checkpoint.acks`: Logstah will checkpoint after this many events are acknowledged. This configuration controls the durability at the processing (filter + output)
+part of Logstash.
+
+Disk writes are have a resource cost. Tuning the above values higher or lower will trade durability for performance. For instance, if you want to the strongest durability for all input events, you can set `queue.checkpoint.writes: 1`.
```
You're right, I should contract it to `are've` an obvious contraction of "are" and "have"..

haha ;)

jordansissel commented:
OK! I think I've resolved all the feedback now.

I've rebased this into a single commit and am going to merge.

Thank you for helping copy edit this important document <3 <3
